### ISSUE TITLE
New rule to prevent reassigning `defaultProps`?

### ISSUE THREAD

**Issue by cangoektas** [yannickcr/eslint-plugin-react#1693]
Hi there,

I just spend about 8 hours of debugging because I accidentally overwrote the `defaultProps` property of a component. 😞 

### Example:
```js
// MyButton.js
const Button = require('./button');

const MyButton = props => {
  return (
    <div className="my-button">
      <Button>{props.text}</Button>
    </div>
  );
};

MyButton.propTypes = {
  text: PropTypes.string.isRequired
};

Button.defaultProps = {}; // <-- Reassigning `defaultProps` of imported component here ⚠️
```

It's just a stupid typo but it did cause a super weird bug for us and tracking this down took a lot of time. I believe there's room for an additional rule. Not sure about the specifics but I would propose something like this:

### Proposal: `no-reassign-default-props`
- doesn't allow reassigning the `defaultProps` property
- doesn't allow assigning the `defaultProps` property for imported components

What do you think?

**Comment by ljharb** [comment]
That seems like it’d be a problem with almost any mutations; perhaps a more generic rule (not react-specific) that prevented mutating imports/requires?

**Comment by cangoektas** [comment]
Hmm yeah, good point. I also have a feeling that those two cases I mentioned should be separated:

1. react-specific rule to prevent reassigning `defaultProps` in the _same file_
2. generic one to prevent cases of mutation for imported modules

Just wondering if the second one is really _always_ problematic... and what cases of mutation should/can we warn about?

**Comment by ljharb** [comment]
Mutating things you don’t own is generally considered always problematic.

**Comment by cangoektas** [comment]
@ljharb do you see something like that as part of `eslint-plugin-import` maybe?

**Comment by ljharb** [comment]
Tbh I’d see it as a core rule; if they reject it, then the import plugin.

**Comment by cangoektas** [comment]
Got you, I'll try to bring that into ESLint... 👍 

What about `no-reassign-default-props` or `no-reassign-react-statics` for mutating static properties of React components? Think that could be useful here?

**Comment by ljharb** [comment]
I don’t think doing it just for defaultProps would be useful; youd also need to do it for propTypes, contextTypes, childContextTypes, and whatever React introduces in the future. A generic rule would be better.
